### ISSUE TITLE
Support HTTPS out of the box

### ISSUE THREAD

**Issue by bajtos** [strongloop/loopback-next#430]
LoopBack.Next applications should support listening on HTTPS out of the box.

An example showing possible usage:

```ts
const app = new Application({
  http: { port: 3000 },
  https: { port: 3001, /* TLS config */ },
  components: [ /*...*/ ],
});

app.start();
// Now we have two servers: HTTP listening on 3000, HTTPS on 3001
```

~Depends on #1331~

## Acceptance criteria

- LB4 app should accept options to configure HTTPS details. It already accepts an options object to configure the REST server. This options should be extended to include HTTPS details. Files affected `@loopback/rest/rest.application.ts`, `@loopback/rest/rest.server.ts`, and `@loopback/http-server`.
- HTTPS configured app should successfully start listening with HTTPS enabled.

Note: We create `port` and `TLS config` keys for https server, other configurations are out of the scope.

**Comment by kjdelisle** [comment]
Quick point-of-note: the constructor should check to make sure that both ports aren't the same.

**Comment by dhmlau** [comment]
@hacksparrow , after discussing with @raymondfeng  and @bajtos , we think we should try to get it in for DP3 (removed the `p2` tag btw).  Could you please add the acceptance criteria for this task, so that we can estimate it on Thu (Jun 14)?  thanks!

**Comment by jannyHou** [comment]
Suggestion from @raymondfeng :
Organize server configurations into a top level property(e.g. transport, or endpoint)

```ts
{
   ...
  endpoints: [{protocol: 'http', port: 3000}, {protocol: 'https', port: 3001}],
  components: []
  // consider to make them an object, so each item can have a key(name)
}
// discriminator: use `protocol`
```

Feel free to add more details in case I missed some important points ^

**Comment by bajtos** [comment]
IIRC, the original plan was to support multiple endpoints by creating multiple RestServer instances, one for each endpoints. Multiple endpoints are out of scope of DP3 though.

As I see the scope of this issue, we want to enhance the current configuration object to allow users to switch from HTTP to HTTPS for the single Rest server their app has.

Originally, I was thinking about the following two flavors of a config object:

HTTP:

```json5
{
  "port": "3000",
  "host": "localhost",
  "https": false //<-- optional
}
```

HTTPS:

```json5
{
  "port": 3000,
  "host": "localhost",
  "https": true,
  "key": "...",
  "cert": "...",
  "passphrase": "...",
   // etc.
}
```

Based on the previous comment, I quite like the idea of using `protocol: 'http'` and `protocol: 'https`'. I find such design very similar to our existing format of datasource configuration, where we use `connector` to discriminate between different databases (storage formats) and the rest of the options are a mix of common things like host/port/username/password and connector-specific options like connection-pool size, database/collection name, etc. It also allows us to extend the list of supported/recognized protocols to other transports like `grpc`, `ws` (WebSockets), `wss` (Secure WebSockets), etc.

Here is my proposal:

HTTP:

```json5
{
  "protocol": "http" //<-- optional, http is default
  "port": "3000",
  "host": "localhost",
}
```

HTTPS:

```json5
{
  "protocol": "https",
  "port": 3000,
  "host": "localhost",
  "key": "...",
  "cert": "...",
  "passphrase": "...",
   // etc.
}
```

Please note that the proposed objects shown above are forward-compatible with the multi-endpoint example - each endpoint entry can be a config object following the convention I proposed.

As for the scope of this user story, we need to:
 - Modify `@loopback/http-server`, `@loopback/rest` and possible other packages to support `"protocol": "https"` and pass the additional options to [`https.createServer`](https://nodejs.org/api/https.html#https_https_createserver_options_requestlistener)
 - Update documentation pages, provide documentation for app developers showing how to enable HTTPS

**Comment by dhmlau** [comment]
Closing as done as PR#1488 has landed.
